flow-server/src/main/java/com/vaadin/flow/server/AppShellRegistry.java

```diff
@@ -211,9 +225,55 @@ private AppShellSettings createSettings() {
             settings.setPageTitle(pageTitles.get(0).value());
         }
         getAnnotations(Inline.class).forEach(settings::addInline);
+
+        Set<String> stylesheets = new LinkedHashSet<>();
+        for (StyleSheet sheet : getAnnotations(StyleSheet.class)) {
+            String href = resolveStyleSheetHref(sheet.value(), contextPath);
+            if (href != null && !href.isBlank()) {
+                stylesheets.add(href);
+            }
+        }
+        stylesheets.forEach(href -> settings.addLink("stylesheet", href));
+
         return settings;
     }
 
+    private String resolveStyleSheetHref(String href, String contextPath) {
+        if (href == null || href.isBlank()) {
+            return null;
+        }
+        href = href.trim();
+        // Accept absolute http(s) URLs unchanged
+        String lower = href.toLowerCase();
+        if (lower.startsWith("http://") || lower.startsWith("https://")) {
+            return href;
+        }
+        // Accept context-relative URLs: context://path -> /path
+        String contextProtocol = ApplicationConstants.CONTEXT_PROTOCOL_PREFIX;
+        if (lower.startsWith(contextProtocol)) {
+            String path = href.substring(contextProtocol.length());
+            if (!path.startsWith("/")) {
+                path = "/" + path;
+            }
+            if (contextPath != null && !contextPath.isEmpty()) {
+                return contextPath + path;
+            }
+            return path;
+        }
+        // Treat ./ as relative path to static resources location
+        if (href.startsWith("./")) {
+            href = href.substring(2);
+        }
+        if (href.contains("..")) {
+            log.warn(
+                    "@StyleSheet href containing traversals ('../') are not allowed, ignored: "
+                            + href);
+            return null;
+        }
+        // Accept bare paths beginning with '/' as-is
+        return href.startsWith("/") ? href : "/" + href;
```

Comment:
```suggestion
        // Accept bare paths beginning with '/' as-is
         href = href.startsWith("/") ? href : "/" + href
         if (HandlerHelper.isPathUnsafe(href)) {
            log.warn(
                    "@StyleSheet href containing traversals ('../') are not allowed, ignored: "
                            + href);
            return null;
        }
        return href;
```
Not 100 that we need this here as these should come from annotations, but lets use the more comprehensive check.

Reply:
Done